web_external/ImagesGallery/ImagesFilters.js

```diff
@@ -0,0 +1,225 @@
+            this._filters = {
+                facetId1: {
+                    binLabel1: true,
+                    binLabel2: false,
+                    ...
+                },
+                facetId2: {
+                    binLabel1: true,
+                    binLabel2: false,
+                    ...
+                },
+                ...
+            }
+        */
+        this._filters = completeFacets.chain()
+            .map(function (completeFacet) {
+                var facetId = completeFacet.id;
+                var facetBins = completeFacet.get('bins');
+
+                var facetFilters = _.chain(facetBins)
+                    .map(function (facetBin) {
+                        var binLabel = facetBin.label;
+                        // Default to all bins included
+                        var binIncluded = true;
+                        return [binLabel, binIncluded];
+                    })
+                    .object()
+                    .value();
+
+                return [facetId, facetFilters];
+            })
+            .object()
+            .value();
+    },
+
+    isIncluded: function (facetId, binLabel) {
+        return this._filters[facetId][binLabel];
+    },
+
+    setIncluded: function (facetId, binLabel, binIncluded) {
+        this._filters[facetId][binLabel] = binIncluded;
+        this.trigger('change:' + facetId);
+        this.trigger('change');
+    },
+
+    setAllIncluded: function (facetId, binsIncluded) {
+        _.each(this._filters[facetId], function (oldValue, binLabel, facetBin) {
+            facetBin[binLabel] = binsIncluded;
+        });
+        this.trigger('change:' + facetId);
+        this.trigger('change');
+    },
+
+    asAst: function () {
+        var fullExpression = _.chain(this._filters)
+            .pick(function (facetFilters, facetId) {
+                // Keep (facetId: facetFilter) elements that have at least one excluded bin
+                return _.chain(facetFilters)
+                    .values()
+                    .contains(false)
+                    .value();
+            })
+            .map(function (facetFilters, facetId) {
+                // Convert each facet into a string filter expression
+                if (_.has(isic.FACET_SCHEMA[facetId], 'lowBound')) {
+                    return this._rangeFacetFilterExpression(facetId, facetFilters);
+                } else {
+                    return this._categoricalFacetFilterExpression(facetId, facetFilters);
+                }
+            }, this)
+            .value()
+            .join(' and ');
+        if (fullExpression) {
+            var ast = isic.SerializeFilterHelpers.astParser.parse(fullExpression);
+            ast = isic.SerializeFilterHelpers._dehexify(ast);
+            return isic.SerializeFilterHelpers._specifyAttrTypes(ast);
+        } else {
+            return undefined;
+        }
+    },
+
+    _rangeFacetFilterExpression: function (facetId, facetFilters) {
```

Comment:
The first 3 issues here were caused by an incorrect handling of null values. See https://github.com/ImageMarkup/isic-archive/pull/365/commits/c9702d496ea787be3356a4295184eeca56055dd8 for a fix.